### PR TITLE
Add `non_blocking=True` for additional train/val loaders

### PR DESCRIPTION
**Pull request by glenn-jocher**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enable non-blocking GPU transfers across training/validation pipelines to improve throughput and reduce data transfer bottlenecks 🚀

### 📊 Key Changes
- Use non-blocking tensor transfers (`.to(device, non_blocking=True)`) for:
  - Classification: images and labels in train/val
  - Detection: `batch_idx`, `cls`, `bboxes` in val
  - Pose: `keypoints` in val
  - Segmentation: `masks` in val
  - YOLO-World: text embeddings in train
  - YOLOE: text embeddings, visuals in train/val
- Minor cleanup: simplified a redundant return in YOLOE trainer

### 🎯 Purpose & Impact
- Faster pipelines: Asynchronous host→device copies can overlap with compute, improving GPU utilization and reducing step time ⚡
- Seamless integration: No API changes; behavior remains the same, with potential speedups when using pinned memory
- Broad coverage: Benefits classification, detection, pose, segmentation, YOLO-World, and YOLOE workflows across train/val
- No accuracy changes: Pure performance optimization; metrics and results remain unchanged

Tip: To maximize benefit, use DataLoader pinned memory and workers, e.g.:
```python
dataloader = torch.utils.data.DataLoader(
    dataset,
    batch_size=bs,
    num_workers=4,
    pin_memory=True,
)
```

### CODE REVIEW

**Comment by glenn-jocher** [comment]
@Laughing-q I think we may have suboptimal device transfers for some train and val loaders, so I've uniformly applied this argument now. Previously was only in the detect trainloader.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! This is an automated response — an Ultralytics engineer will review and assist soon. In the meantime, please review the checklist below to help streamline the process:

- ✅ **Define a Purpose**: Confirm the PR description clearly states the goal of adding `non_blocking=True` to additional train/val loaders and links to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure commit messages are clean and follow project conventions.
- ✅ **Synchronize with Source**: Ensure your branch is up-to-date with the `ultralytics/ultralytics` `main` branch. If it’s behind, click ‘Update branch’ or run `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are green. If anything fails, please address it.
- ✅ **Update Documentation**: If applicable, update the relevant [documentation](https://docs.ultralytics.com/) to reflect the non-blocking transfers and any best practices (e.g., pinned memory).
- ✅ **Add Tests**: Where relevant, include or update tests across classification, detection, segmentation, pose, WORLD, and YOLOE to cover the data transfer changes, and confirm all tests pass locally and in CI.
- ✅ **Sign the CLA**: If this is your first Ultralytics PR, please sign the [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) by commenting: "I have read the CLA Document and I sign the CLA".
- ✅ **Minimize Changes**: Keep modifications focused on the intended non-blocking device transfers and small cleanups. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

Additional notes for this PR:
- 👍 Thanks for covering classification, detection, segmentation, pose, WORLD, and YOLOE paths. Please double-check consistency of `.to(..., non_blocking=True)` usage across train/val to avoid regressions.
- 🧪 If you have quick benchmarks demonstrating the speedups (especially with `pin_memory=True` and multiple workers), consider sharing in the PR description to help reviewers contextualize impact.

For more guidance, please see our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Thanks for improving Ultralytics! ⚡

**Review by Laughing-q — APPROVED** [review]

**Comment by UltralyticsAssistant** [comment]
Thrilled to see this merged! Huge thanks to @glenn-jocher and kudos to @Laughing-q for the thoughtful contributions in [PR #21912](https://github.com/ultralytics/ultralytics/pull/21912) 🚀

“Efficiency is doing better what is already being done.” — Peter Drucker. Your non-blocking GPU transfers embody this perfectly—seamlessly boosting throughput across classification, detection, pose, segmentation, YOLO-World, and YOLOE without changing user workflows. It’s the kind of behind-the-scenes optimization that keeps Ultralytics models running faster and smoother for everyone. Appreciate the precision and care here!
